File: packages/shared/src/types/job.ts
/**
 * Job models - represents background processing jobs
 */

export type JobType =
  | 'terrain_generation'
  | 'seed_placement'
  | 'growth_simulation'
  | 'vmt_calculation'
  | 'export_png'
  | 'export_gif'

export type JobStatus =
  | 'pending'
  | 'running'
  | 'completed'
  | 'failed'
  | 'cancelled'

export interface JobCreate {
  type: JobType
  /** Target tile (if applicable) */
  tile_id?: string | null
  /** Job-specific parameters */
  params?: Record<string, unknown>
}

export interface Job {
  id: string
  type: JobType
  status: JobStatus
  tile_id: string | null
  params: Record<string, unknown>
  result: Record<string, unknown> | null
  error: string | null
  created_at: string
  started_at: string | null
  completed_at: string | null
}

File: apps/api/src/city_api/schemas/job.py
"""Job schemas - represents background processing jobs."""

from datetime import datetime
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class JobType(StrEnum):
    """Types of background jobs."""

    TERRAIN_GENERATION = "terrain_generation"
    SEED_PLACEMENT = "seed_placement"
    GROWTH_SIMULATION = "growth_simulation"
    VMT_CALCULATION = "vmt_calculation"
    CITY_GROWTH = "city_growth"
    EXPORT_PNG = "export_png"
    EXPORT_GIF = "export_gif"


class JobStatus(StrEnum):
    """Status of a background job."""

    PENDING = "pending"
    RUNNING = "running"
    CLAIMED = "claimed"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class JobCreate(BaseModel):
    """Request model for creating a new job."""

    type: JobType
    tile_id: UUID | None = Field(
        default=None,
        description="Target tile (if applicable)",
    )
    params: dict[str, Any] = Field(
        default_factory=dict,
        description="Job-specific parameters",
    )


class Job(BaseModel):
    """A background processing job."""

    id: UUID
    user_id: UUID
    type: JobType
    status: JobStatus = JobStatus.PENDING
    tile_id: UUID | None = None
    params: dict[str, Any] = Field(default_factory=dict)
    result: dict[str, Any] | None = None
    error: str | None = None
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None

    model_config = {"from_attributes": True}
